### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Ugboaja-Uchechi**
###  Project Requirements

- Make sure YAGNI, KISS, and DRY rules and Ruby style guide are respected.

- Follow SOLID principles.

- Pay special attention to the "Single-responsiblity Principle".

- Make sure that the role of main.rb is only to start your code and ask the user for the initial choice of options (or to quit).

- Extract some pieces of logic into a separate class(es).

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    when '5'
+      create_rental
+    when '6'
+      rentals_by_id
+    else
+      puts 'Wrong option'
+      puts '--------------------------------------'
+      puts 'Please enter a number between 1 and 7'
+    end
+  end
```
Comment (Mihndim2020) [inline app.rb:35]: - [x] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    print 'Name: '
+    name = gets.chomp
+    case input
+    when 1
+      create_student(age, name)
+    when 2
+      create_teacher(age, name)
+    end
+    puts 'Person created successfully!'
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [x] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    print 'Has parent permission? [Y/N]: '
+    parent_permission = gets.chomp.upcase == 'Y'
+    @people << Student.new(name, age, parent_permission)
+  end
+
+  def create_teacher(age, name)
+    print 'Specialization: '
+    specialization = gets.chomp
+    @people << Teacher.new(name, age, specialization)
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [x] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    print 'Author: '
+    author = gets.chomp
+
+    book = Book.new(title, author)
+    @books << book
+
+    puts 'Book added successfully'
+    sleep 1
+    puts
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [ ] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    end
+    person_index = gets.chomp.to_i
+    puts person_index
+    print 'Date: '
+    date = gets.chomp
+    @rentals << Rental.new(date, @people[person_index], @books[book_index])
+    puts 'Rental created successfully'
+    sleep 1
+    puts
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [x] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  def rentals_by_id
+    print 'ID of person: '
+    id = gets.chomp.to_i
+    puts 'Rentals:'
+    @rentals.each do |rental|
+      puts "Date: #{rental.date}, Book \"#{rental.book.title}\" by #{rental.book.author}" if rental.person.id == id
+    end
+    sleep 1
+    puts
+  end
```
Comment (Mihndim2020) [inline app.rb]: - [x] Following the `Single responsibility rule`, all requests for input and input validation should placed in classes/methods whose responsibilities are to request or validate input. Please checkout this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) and kindly fix that.

**Review by Mihndim2020 — CHANGES_REQUESTED** [review]
### Hi Team, 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages,
:heavy_check_mark: Descriptive Readme file,
:heavy_check_mark: Your app is running free of `Linter` errors.

There are some issues that you still need to work on to get this project approved, but you are almost there!
ALMOST-MINION

# Status: Requires Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      puts "#{index}) [#{person.class}] Name: #{person.name}, ID: #{person.id}, Age: #{person.age}"
+    end
+    person_index = user_input('', 'int')
+    puts person_index
+    date = user_input('Date: ', 'str')
+    @rentals << Rental.new(date, @people[person_index], @books[book_index])
+    puts 'Rental created successfully'
+    sleep 1
+    puts
+  end
```
Comment (mikethreels) [inline app.rb:102]: - [ ]  in these methods you are printing to the console, getting information and saving the data which is not in line with the single responsibility rule, please have a look at [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) for more information. try to separate the output and input to and from the user from saving your data.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request closed without merging**